### DISCUSSION TITLE
Can you not use an inherited `__table__` wrapped with `declared_attr` from a base class in 2.x?

### DISCUSSION THREAD

**Discussion by trogers8080** [sqlalchemy/sqlalchemy#11508] [Usage Questions] [answered]
I'm upgrading our models from 1.4 to 2.0 and it mostly has gone smoothly. However we have several models that represent views in our db mapped using a hybrid declarative style. These view models inherit a `__table__` attribute from a base class that has a `__table__`  declared as a `declared_attr` function. After upgrading to 2.0, an error is thrown saying the models don't have a `__tablename__` or `__table__` defined. Is this not supported anymore? I dug through the source for the declarative process and it seems it explicitly skips inheriting a `__table__` and only looks for it in the class's `__dict__`.

For extra context, the base class's `__table__` `declared_attr` function constructs and returns table object from a sqlalchemy query using [sqlalchemy-utils's `create_view()`](https://sqlalchemy-utils.readthedocs.io/en/latest/view.html#create-view) function. The query is provided by the individual subclass like so:
```
reg = registry()

class ViewBase(DeclarativeBase):
    registry = reg

    @classmethod
    def _get_query(cls):
        raise NotImplementedError("")

    @declared_attr
    def __table__(cls) :
        return create_view(cls.__name__, cls.get_query(), cls.registry.metadata)

class ViewInDb(ViewBase):
    @classmethod
    def _get_query(cls):
        return select(TableInDb.col_a, TableInDb.col_b).where(TableInDb.col_a == "foo")
```

This code was working in 1.4 but after upgrading   to 2.0 gives me the above mentioned error complaining that `ViewInDb` doesn't have a `__tablename__` nor a `__table__` defined. I wasn't sure if this was a bug or not since I don't see anything in the change log talking about this change, and the docs don't make it clear whether you can or cannot do this.

**Comment by zzzeek** [comment, marked as answer]
> After upgrading to 2.0, an error is thrown saying the models don't have a __tablename__ or __table__ defined. Is this not supported anymore? 

this is supported.   The behavior is a bug #11509 which will be fixed by tonight.   Place `__table__` on a subclass of `ViewBase` declared with `__abstract__ = True` for now.

**Reply by zzzeek** [reply]
abstract seems not not work either.  you might be able to get away with ``__table_cls__()``

**Reply by trogers8080** [reply]
Thanks for the quick response! I tried the `__table_cls__()` way but that gave me the same error as above. I ended up working around it using `__init_subclass__()` on the base class to set ` cls.__table__` before calling the `DeclarativeBase`'s `__init_subclass__()`.

You said it will be fixed by tonight. Does that mean it will be fixed with a new version in pypi tonight or just fixed in the codebase tonight but deployed to pypi at a later time? If the latter, do you have any idea when that deploy will happen?

**Reply by zzzeek** [reply]
it means there's a fix in review that will be in the next release, we're due for next release so should be this week.

you can assign ``__table__`` directly on each subclass to work around:

```py
from __future__ import annotations

from sqlalchemy import Column, select
from sqlalchemy import Integer
from sqlalchemy import Table
from sqlalchemy.orm import DeclarativeBase
from sqlalchemy.orm import declared_attr


class BaseWithTable(DeclarativeBase):

    def _make_a_table(cls):
        return Table(
            cls.__name__,
            cls.registry.metadata,
            Column("id", Integer, primary_key=True),
        )


class HasTable(BaseWithTable):
    __table__ = declared_attr(BaseWithTable._make_a_table)


class AlsoHasATable(BaseWithTable):
    __table__ = declared_attr(BaseWithTable._make_a_table)


print(select(HasTable))

```
